### PR TITLE
Implemented pixel inefficiency at ROC level

### PR DESCRIPTION
**Pull request by tsusa**
The PR extends the functionality of the pixel inefficiency simulation (SiPixelDynamicInefficiency) in digitizer from the module to the ROC level. It was needed to simulate the damage in ROCs caused by the DCDC issue. The damage is bigger for the modules operated longer in such conditions (LV off/HV on), effect is also larger for ROCs closer to the beam.
The damage first appears in big pixels and then the standard size pixels. For that reason, the first 5% (percentage of the big pixels in the ROC) of bad pixels is assigned to the big ones, and the remaining part to the standard pixels.

As SiPixelDynamicInefficiency DB object has not been updated yet, no changes expected with this PR.

The plots showing pixelAlive measurements and results of the simulation for several modules are attached below. 
[PR_bpix.pdf](https://github.com/cms-sw/cmssw/files/1782047/PR_bpix.pdf)
[PR_fpix.pdf](https://github.com/cms-sw/cmssw/files/1782048/PR_fpix.pdf)

@veszpv @tvami @boudoul @jkarancs @ferencek

### CODE REVIEW

File: SimTracker/SiPixelDigitizer/plugins/SiPixelDigitizerAlgorithm.cc [hunk 1]
```diff
@@ -496,27 +496,62 @@ void SiPixelDigitizerAlgorithm::init_DynIneffDB(const edm::EventSetup& es, const
...
     ColGeomFactors[rawid] = 1;
     ChipGeomFactors[rawid] = 1;
+    PixelGeomFactorsROCStdPixels[rawid] = std::vector<double>(16,1);
+    PixelGeomFactorsROCBigPixels[rawid] = std::vector<double>(16,1);
+  }
+  
+  // Store ROC level scale factors  (big and standard pixels separatelly)
+  std::map<uint32_t, double>  PixelGeomFactorsDB;
+  for (auto db_factor : PixelGeomFactorsDBIn){  
+    int shift = DetId(db_factor.first).subdetId() == 1? 6 : 3;
```
Comment (veszpv) [inline SimTracker/SiPixelDigitizer/plugins/SiPixelDigitizerAlgorithm.cc]: Dear Tanja, could you perhaps 1) use the enum for the condition on the subdetId() like PixelSubdetector::PixelBarrel or PixelSubdetector::PixelEndcap, 2) define some enum or similar for the respective ROC id mask and shift, and 3) make sure you are not trying to use the same bits for Phase 0 (or turn on this feature at all). In Phase 0 BPix, module coordinates had only 6 bits allocated in the DetId.

Comment (veszpv) [reply SimTracker/SiPixelDigitizer/plugins/SiPixelDigitizerAlgorithm.cc]: btw, you have access to the TrackerGeometry below. In principle, geom->getDetectorType() should tell is the code is running on Phase 0 or Phase 1

File: SimTracker/SiPixelDigitizer/plugins/SiPixelDigitizerAlgorithm.cc [hunk 2]
```diff
@@ -496,27 +496,62 @@ void SiPixelDigitizerAlgorithm::init_DynIneffDB(const edm::EventSetup& es, const
...
+    unsigned int rocMask = 0x1F << shift;
+    unsigned int rocId = (((db_factor.first) & rocMask) >> shift);
+    if (rocId != 0) {
+      rocId--;
+      unsigned int rawid = db_factor.first & (~rocMask);      
+      double factor = db_factor.second;
+      double badFraction = 1 - factor;
+      double bigPixelFraction = 210./4160.;      
+      double stdPixelFraction = 3950./4160.;      
+
```
Comment (veszpv) [inline SimTracker/SiPixelDigitizer/plugins/SiPixelDigitizerAlgorithm.cc]: These numbers can be obtained by doing something like this somewhere where you have a concrete pixel DetId, or already a PixelGeomDetUnit:
const PixelGeomDetUnit * theGeomDet = dynamic_cast<const PixelGeomDetUnit*> (geom->idToDet(...a pixel det ID here...));
PixelTopology const * topology = &(theGeomDet->specificTopology());
and using topology->rowsperroc(), colsperroc(), nrows(), and ncolumns()

File: SimTracker/SiPixelDigitizer/plugins/SiPixelDigitizerAlgorithm.cc [hunk 3]
```diff
@@ -1579,6 +1620,10 @@ void SiPixelDigitizerAlgorithm::pixel_inefficiency(const PixelEfficiencies& eff,
     
     chips[chipIndex]++;
     columns[dColInDet]++;
+    if (row == 79 || row == 80 || col % 52 == 0 || col % 52 == 51)
```
Comment (veszpv) [inline SimTracker/SiPixelDigitizer/plugins/SiPixelDigitizerAlgorithm.cc]: I think this should work for both Phase 0 and 1 if the feature is not turned off for Phase 0. (Although in Phase 0 we had plaquettes where row 80 did not exist.) Again, rowsperroc() and colsperroc() would look better.

**Comment by makortel** [comment]
Do I guess correctly that in the context of premixing these code changes affect premixing stage2 (=overlay of the premixed pileup event)? With the current code organization these would then have to be transferred (copy-pasted) to [SimGeneral/DataMixingModule/plugins/DataMixingSiPixelMCDigiWorker.cc](https://github.com/cms-sw/cmssw/blob/master/SimGeneral/DataMixingModule/plugins/DataMixingSiPixelMCDigiWorker.cc) (or `SimGeneral/PreMixingModule/plugins/PreMixingSiPixelWorker.cc` after #22416 gets merged)

(and we should try to remove the code duplication)

@mdhildreth @kpedro88

**Comment by tsusa** [comment]
@makortel 
yes the code should be transferred to  SimGeneral/DataMixingModule/plugins/DataMixingSiPixelMCDigiWorker.cc

**Comment by makortel** [comment]
Thanks @tsusa. I guess we have two options to proceed
1. Transfer the code in this PR (and create conflict with #22416, straightforward to resolve though so not a big deal)
2. Merge this PR and #22416 separately, and then create a third PR transferring the code (no conflicts but then there is a time window when classical mixing and premixing are out of synch; OTOH the functionality of this PR is not enabled yet)

@fabiocos @davidlange6 Do you have a preference?

**Comment by tsusa** [comment]
@makortel  Has any decision been made concerning the transfer of the code to the  DataMixingModule/PreMixing Module?

**Comment by makortel** [comment]
@tsusa I'm not aware of any explicit decision, I guess it is up to @fabiocos @davidlange6 @civanch @mdhildreth etc to comment.

#22416 is planned to be merged for 10_2, so if the code ought to be transferred in this PR and this PR ought to be merged in 10_1, then it must go to `DataMixingSiPixelMCDigiWorker.cc` (and I'll deal with it in #22416).

**Comment by tsusa** [comment]
@makortel  thanks and sorry, I was not sure who should be asked, Yes, for us it  would be important to have this PR merged in 10_1

**Comment by fabiocos** [comment]
@tsusa sorry, this is clearly a development for 10_2_X, we should not touch simulation any more in 10_1_X. I assume your target is performing tests, right?

**Comment by tsusa** [comment]
@fabiocos no, we were targeting 10_1_X MC for POG studies

**Comment by tsusa** [comment]
@fabiocos please have a look at 
https://indico.cern.ch/event/711247/contributions/2927858/attachments/1614749/2565773/tanjaReco_09032018.pdf, slide 9 and https://twiki.cern.ch/twiki/bin/viewauth/CMS/CMSSW_10_1_0

**Comment by civanch** [comment]
please test

**Comment by civanch** [comment]
@tsusa , number of differences substantial. Is this expected?

**Comment by tsusa** [comment]
@civanch We do not expect the differences if the same GT is used. From the link above it is not easy to find which quantities have been affected.  Could you please point me to some examples?

**Comment by makortel** [comment]
@tsusa E.g. here (from workflow 10024, i.e. 2017 ttbar noPU)
https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_10_1_X_2018-03-25-2300+22465/25835/10024.0_TTbar_13+TTbar_13TeV_TuneCUETP8M1_2017_GenSimFull+DigiFull_2017+RecoFull_2017+ALCAFull_2017+HARVESTFull_2017/PixelPhase1_Phase1_MechanicalView_PXBarrel_Shell_mI.html
https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_10_1_X_2018-03-25-2300+22465/25835/10024.0_TTbar_13+TTbar_13TeV_TuneCUETP8M1_2017_GenSimFull+DigiFull_2017+RecoFull_2017+ALCAFull_2017+HARVESTFull_2017/PixelPhase1_Phase1_MechanicalView_PXForward_HalfCylinder_mI.html
https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_10_1_X_2018-03-25-2300+22465/25835/10024.0_TTbar_13+TTbar_13TeV_TuneCUETP8M1_2017_GenSimFull+DigiFull_2017+RecoFull_2017+ALCAFull_2017+HARVESTFull_2017/Tracking_TrackParameters_generalTracks_GeneralProperties.html

**Comment by tsusa** [comment]
Thanks @markotel, but the links to the comparison plots got broken

**Comment by boudoul** [comment]
someone should retrigger the tests...( I don't have this power )

**Comment by kpedro88** [comment]
please test

**Comment by tsusa** [comment]
sorry, not old but the current payload

**Comment by tsusa** [comment]
How should we proceed?

**Comment by civanch** [comment]
+1

File: SimTracker/SiPixelDigitizer/plugins/SiPixelDigitizerAlgorithm.cc [hunk 4]
```diff
@@ -496,27 +496,74 @@ void SiPixelDigitizerAlgorithm::init_DynIneffDB(const edm::EventSetup& es, const
...
-  // Loop on all modules, calculate geometrical scale factors and store in map for easy access
+  // Loop on all modules, initialize map for easy access
   for( const auto& it_module : geom->detUnits()) {
     if( dynamic_cast<PixelGeomDetUnit const*>(it_module)==nullptr) continue;
     const DetId detid = it_module->geographicalId();
     uint32_t rawid = detid.rawId();
     PixelGeomFactors[rawid] = 1;
     ColGeomFactors[rawid] = 1;
     ChipGeomFactors[rawid] = 1;
+    PixelGeomFactorsROCStdPixels[rawid] = std::vector<double>(16,1);
```
Comment (fabiocos) [inline SimTracker/SiPixelDigitizer/plugins/SiPixelDigitizerAlgorithm.cc]: was the memory impact of these maps evaluated? Which is the extra memory taken by this module after the update?

Comment (tsusa) [reply SimTracker/SiPixelDigitizer/plugins/SiPixelDigitizerAlgorithm.cc]: The size of maps is ~ # of modules x # number of ROCs x sizeof(double) = 1856 x 16 x 8  = 0.24 MB

**Comment by fabiocos** [comment]
+1 

on a test run on 100 events of wf 10824 no significant memory impact is observed

**Comment by makortel** [comment]
Now that this feature seems to aim for production
https://hypernews.cern.ch/HyperNews/CMS/get/relval/10650.html
should we do something for the premixing implementation (see the options in my comment above) before discovering discrepancies in premixing vs. classical mixing comparisons? (IIUC the current validation campaign does not include premixing)

**Comment by makortel** [comment]
Just to note that #23556 reduces the code duplication between `PreMixingSiPixelWorker` and `SiPixelDigitizerAlgorithm` so that with that PR the feature of this PR should be automatically included in premixing (stage2).

**Pull request merged**